pytorch_ops/generate_tests.py

```diff
@@ -0,0 +1,201 @@
+            print("", file=config)
+
+    def generate_test(self, rtol=1e-05, atol=1e-08, equal_nan=False):
+        """
+        The default values for rtol, atol, and equal_nan come from the
+        numpy.allclose's default values listed in the link below.
+        https://numpy.org/devdocs/reference/generated/numpy.allclose.html
+
+        These values will be saved in the json file and be used when
+        running the test.
+        """
+        inputs = self.generate_inputs()
+        self.test_config["rtol"] = rtol
+        self.test_config["atol"] = atol
+        self.test_config["equal_nan"] = equal_nan
+        self.save_mlir(*inputs)
+        expected_results = self.generate_expected_value(*inputs)
+        self.save_inputs(*inputs)
+        self.save_results(*expected_results)
+        self.save_config()
+
+    def generate_expected_value(self, *args):
+        results = self.forward(*args)
+        return [results]
+
+    def generate_inputs(self):
+        return self.args
+
+    @abstractmethod
+    def forward(self, *args):
+        ...
+
+
+@dataclass
+class RandomSession:
+    generator: torch.Generator = torch.Generator()
+    seed: int = 0
+
+    def __post_init__(self):
+        self.generator.manual_seed(self.seed)
+
+    def rand(self, *args, **kwargs):
+        if kwargs.get("generator"):
+            raise ValueError("RandomSession uses its own generator")
+        kwargs["generator"] = self.generator
+        return torch.rand(*args, **kwargs)
+
+
+class AB(TestGenerator):
+    def forward(self, left, right):
+        return left @ right
+
+
+class AB_bfloat16(TestGenerator):
+    def forward(self, left, right):
+        left = left.to(torch.bfloat16)
+        right = right.to(torch.bfloat16)
+        res = left @ right
+        return res.to(torch.float32)
+
+
+class ATB(TestGenerator):
+    def forward(self, left, right):
+        return left.t() @ right
+
+
+class ABT(TestGenerator):
+    def forward(self, left, right):
+        return left @ right.t()
+
+
+class ABplusC(TestGenerator):
+    def forward(self, A, B, C):
+        return A @ B + C
+
+
+class ReluABPlusC(TestGenerator):
+    def forward(self, A, B, C):
+        return torch.relu(A @ B + C)
+
+
+class GeluABPlusC(TestGenerator):
+    def forward(self, A, B, C):
+        return torch.ops.aten.gelu.default(A @ B + C)
+
+
+# TODO: Future PR, itertools.product where it makes sense
+for dtype in [torch.float32]:
+    for cls in [AB_bfloat16]:
+        random = RandomSession()
+        inputs = (random.rand(64, 64, dtype=dtype), random.rand(64, 64, dtype=dtype))
+        dyn_dim = torch.export.Dim("N")
+        dynamic_shapes = {"left": {0: dyn_dim}, "right": {1: dyn_dim}}
+        instance = cls(
+            *inputs,
+            name=cls.__name__,
+            export_kwargs={"dynamic_shapes": dynamic_shapes},
+        )
+        instance.generate_test(atol=1e-2, rtol=1e-2)
+for dtype in [torch.float32, torch.float16]:
+    for cls in [AB]:
```

Comment:
These loops are not very scalable. I would prefer if we made each class self contained. A class definition should generate a test on it's own.

You can ask each class to have a seed to use to generate something. I'd say do it pytest style where every  marked function is a test. Something like:

```python
class Matmul(TestGenerator):
  
  @test(seed=42)
  def one(self, seed):
    lhs = torch.rand(64, 64, dtype=dtype)
    rhs = torch.randn(64, 64, dty

  def forward(self, lhs, rhs):
    return lhs @ rhs

  def dynamic_shapes(self):
    dyn_dim = torch.export.Dim("N")
    return {
      "left": {0: dyn_dim},
      "right": {1: dyn_dim},
    }
```